### PR TITLE
Feature/detect object

### PR DESCRIPTION
**Pull request by Derstilon**
image
Todo:
- [x] Add segments density to configuration
- [x] Visualize segment density
- [x] Hide material configuration
- [x] Switch to src/ThreeEditor/util/UiUtils.js functions

### CODE REVIEW

**Comment by grzanka** [comment]
@Derstilon can we orient cylinder-like scorers in such way that the symmetry axis is oriented along Z ?

**Comment by grzanka** [comment]
@Derstilon how to define on how many slices detector is divided ?

**Comment by Derstilon** [comment]
> @Derstilon how to define on how many slices detector is divided ?

This will be added soonish

**Comment by grzanka** [comment]
> > @Derstilon how to define on how many slices detector is divided ?
> 
> This will be added soonish

Will it be in this PR or in a new one ?

File: src/ThreeEditor/util/Detect/DetectTypes.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    heightSegments: number
+};
+
+export type Zone = {
+    zoneId: number
+};
+
+export type All = {};
+
+export type Any = {
```
Comment (ostatni5) [inline src/ThreeEditor/util/Detect/DetectTypes.ts:23]: https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#intersection-types

**Comment by Derstilon** [comment]
> > > @Derstilon how to define on how many slices detector is divided ?
> > 
> > 
> > This will be added soonish
> 
> Will it be in this PR or in a new one ?

All checkboxes has been marked in the latest commit. "detector segments feature" was included in one of them.

**Comment by grzanka** [comment]
Currently it is possible to set number of bins to non-integer values.
Can we change that, so that only integers greater or equal 1 are allowed ?
image

**Comment by grzanka** [comment]
Please set default bins in radius to 1, and increase number of bins along Z to 100

image

**Comment by grzanka** [comment]
Cylyndrical mesh is not being displayed

image

File: src/ThreeEditor/js/Sidebar.Geometry.Detect.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	const container = new UISpan();
+	let type = section.detectGeometryType;
+	let data = section.detectGeometryData;
+
+	// detect type
+	const [sectionTypeRow, sectionType] = createRowSelect({ text: `Type`, update: updateType, options: detectOptions });
+
+	// width 
+	const [widthRow, width] = createRowParamNumber({ text: `X side length (width) ${editor.unit.name}`, min: 0, update });
+	const [widthRow2, width2] = createRowParamNumber({ text: `X side density (segments)`, min: 1, update, precision: 1 });
```
Comment (grzanka) [inline src/ThreeEditor/js/Sidebar.Geometry.Detect.js:16]:
```suggestion
	const [widthRow2, width2] = createRowParamNumber({ text: `number of bins along X`, min: 1, update, precision: 1 });
```

File: src/ThreeEditor/js/Sidebar.Geometry.Detect.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+	// depth
+	const [depthRow, depth] = createRowParamNumber({ text: `Z side length (depth) ${editor.unit.name}`, min: 0, update });
+	const [depthRow2, depth2] = createRowParamNumber({ text: `Z side density (segments)`, min: 1, update, precision: 1 });
+	depth2.setNudge(1).setStep(1);
+
+	// radius
+	const [radiusRow1, radius1] = createRowParamNumber({ text: `Inner radius ${editor.unit.name}`, min: 0, update });
+	const [radiusRow2, radius2] = createRowParamNumber({ text: `Outer radius ${editor.unit.name}`, min: 0, update });
+	const [radiusRow3, radius3] = createRowParamNumber({ text: `Radial density (segments)`, min: 1, update, precision: 1 });
```
Comment (grzanka) [inline src/ThreeEditor/js/Sidebar.Geometry.Detect.js:32]:
```suggestion
	const [radiusRow3, radius3] = createRowParamNumber({ text: `Number of bins along radial axis`, min: 1, update, precision: 1 });
```

File: src/ThreeEditor/js/Sidebar.Geometry.js [hunk 4]
```diff
@@ -61,62 +64,58 @@ function SidebarGeometry(editor) {
...
-
-			const object = editor.selected;
+	async function build() {
+		
+		const object = editor.selected;
 
+		if (isZone(object)) {
 			parameters.clear();
-
+			let zonePanel = new UISpan();
```
Comment (ostatni5) [inline src/ThreeEditor/js/Sidebar.Geometry.js:73]:
```suggestion
			const zonePanel = new UISpan();
```

File: src/ThreeEditor/js/Sidebar.Geometry.js [hunk 5]
```diff
@@ -161,9 +160,11 @@ function SidebarGeometry(editor) {
...
 		currentGeometryType = null;
 		vertexNormalsButton.setDisplay('block');
-		(editor.selected?.isCSGZone
-			? buildZonesManager
-			: build)();
+		build()
+		// (isZone(editor.selected)
+		// 	? buildZonesManager
+		// 	: build
+		// )();
```
Comment (ostatni5) [inline src/ThreeEditor/js/Sidebar.Geometry.js:167]: Remove

File: src/ThreeEditor/js/commands/AddObjectCommand.js [hunk 6]
```diff
@@ -39,7 +39,7 @@ class AddObjectCommand extends Command {
 
 	toJSON() {
 
-		const output = super.toJSON( this );
+		let output = super.toJSON( this );
```
Comment (ostatni5) [inline src/ThreeEditor/js/commands/AddObjectCommand.js:42]:
```suggestion
		const output = super.toJSON( this );
```

File: src/ThreeEditor/js/commands/RemoveZoneCommand.js [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { Command } from '../Command.js';
 
-import { CSGZone } from '../../util/CSG/CSGZone';
+import { Zone } from '../../util/CSG/CSGZone';
```
Comment (ostatni5) [inline src/ThreeEditor/js/commands/RemoveZoneCommand.js:3]: import * as CSG

File: src/ThreeEditor/js/commands/SetDetectGeometryCommand.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+        super(editor);
+
+        this.type = 'SetDetectGeometryCommand';
+        this.name = 'Set DetectGeometry';
+        this.updatable = true;
+
+        this.object = object;
+        this.oldData = object.getData();
+        console.log(newData,this.oldData);
```
Comment (ostatni5) [inline src/ThreeEditor/js/commands/SetDetectGeometryCommand.js:22]:
```suggestion
```

File: src/ThreeEditor/util/Beam.ts [hunk 9]
```diff
@@ -125,24 +155,25 @@ export class Beam extends THREE.Object3D implements ISimulationObject {
...
+        ) as this;
     }
 
     toJSON() {
         const jsonObject: BeamJSON = {
-            type: this.type,
             position: this.position.toArray(),
             direction: this.direction.toArray(),
-            energy: this.energy
+            energy: this.energy,
```
Comment (ostatni5) [inline src/ThreeEditor/util/Beam.ts:176]:
```suggestion
            energy: this.energy
```

File: src/ThreeEditor/util/BoundingZone.ts [hunk 10]
```diff
@@ -49,10 +53,25 @@ export class BoundingZone extends THREE.Object3D implements ISimulationObject {
...
+                switch(p){
+                    case "notMovable":
+                        result = this.autoCalculate && this.canCalculate();
+                        break;
+                    default:
+                        result = Reflect.get(target, p);
+                }
+            return result;
+        },
+    };
```
Comment (ostatni5) [inline src/ThreeEditor/util/BoundingZone.ts:73]: Move to getter

File: src/ThreeEditor/util/Detect/DetectTypes.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        innerRadius: true,
+        depth: true,
+        radialSegments: true,
+        depthSegments: true
+    },
+    Zone:{
+        zoneUuid: true
+    },
+    All:{}
+}
```
Comment (ostatni5) [inline src/ThreeEditor/util/Detect/DetectTypes.ts]:
```suggestion
```

**Review by ostatni5 — APPROVED** [review]
Good Job!
obraz

**Comment by grzanka** [comment]
@Derstilon fix conflicts

**Comment by grzanka** [comment]
@Derstilon fix conflicts

**Review by ostatni5 — APPROVED** [review]
Good job
obraz

**Pull request merged**